Implement RSSReport class

gita-omr opened the pull request:
- Implement a class that can monitor RSS status of any number of arbitry memory regions
- In addition to start address and size, region can have name, as well as a list of items describing its content

gita-omr commented:
Fixed white spaces.

gita-omr commented:
Removed Options.hpp include since it's not needed.

gita-omr commented:
Build fixes.

gita-omr commented:
Fix for windows build.

gita-omr commented:
Another Windows build fix.

gita-omr commented:
More Windows build fixes.

gita-omr commented:
Another Windows warning.

gita-omr commented:
another warning

gita-omr commented:
Corrected second commit

gita-omr commented:
I will fix line endings together with PR comments.

gita-omr commented:
Fixed Mac build and line endings.

gita-omr commented:
FYI: @mpirvu @tajila 

Note: further improvements are possible:
1) options for reporting only specific types of memory (e.g. data cache only, code cache only etc.)
2) disclaiming regions from within the class

dsouzai commented on compiler/runtime/OMRRSSReport.cpp:
```diff
@@ -0,0 +1,354 @@
+   {
+   size_t totalRSSKb = 0;
+   int32_t regionNum = 0;
+   TR::CodeCache *codeCache = TR::CodeCacheManager::instance()->getFirstCodeCache();
```
This appears to be unused.

gita-omr replied:
Right, removed.

dsouzai commented on compiler/runtime/OMRRSSReport.cpp:
```diff
@@ -0,0 +1,354 @@
+OMR::RSSReport::printTitle()
+   {
+   int32_t regionNum = 0;
+   char rssLine[MAX_RSS_LINE];
```
Wouldn't it be more readable if the regions were printed out sequentially, rather them being all on one line?

gita-omr replied:
If we want some frequent statistics it's much more readable in one line. For example, we might get this type of repport:

```
                                   cold cache cold cache
#PERF:  RSS Report(2 regions):   148(100%)  576(100%)  totalRSS=2896KB
#PERF:  RSS Report(2 regions):   148(100%)  576(100%)  totalRSS=2896KB
#PERF:  RSS Report(2 regions):   148(100%)  576(100%)  totalRSS=2896KB   
#PERF:  RSS Report(2 regions):     6(  4%)   72( 12%)  totalRSS=312KB   
#PERF:  RSS Report(2 regions):    35( 24%)  257( 45%)  totalRSS=1168KB 
#PERF:  RSS Report(2 regions):    45( 30%)  291( 50%)  totalRSS=1344KB
#PERF:  RSS Report(2 regions):    51( 34%)  306( 53%)  totalRSS=1428KB
#PERF:  RSS Report(2 regions):    54( 36%)  314( 54%)  totalRSS=1472KB
```
In other words, it's easier to see the changes in RSS for each region when printed in columns.

gita-omr replied:
Also please see https://github.com/eclipse/omr/pull/7349#issuecomment-2135770954

dsouzai commented on compiler/runtime/OMRCodeCacheManager.hpp:
```diff
@@ -30,6 +30,7 @@
 #include "runtime/MethodExceptionData.hpp"
 #include "runtime/Runtime.hpp"
 #include "runtime/CodeCacheTypes.hpp"
+#include "runtime/OMRRSSReport.hpp"
```
Is this needed here?

gita-omr replied:
Removed.

dsouzai commented on compiler/x/codegen/OMRCodeGenerator.cpp:
```diff
@@ -1874,6 +1875,79 @@ struct DescendingSortX86DataSnippetByDataSize
    };
+
+
+void OMR::X86::CodeGenerator::addItemsToRSSReport(uint8_t *oldColdCode, uint8_t *coldCode)
```
Doesn't look like `oldColdCode` is used here.

gita-omr replied:
Right! Removed.

dsouzai commented on compiler/x/codegen/OMRCodeGenerator.cpp:
```diff
@@ -1874,6 +1875,79 @@ struct DescendingSortX86DataSnippetByDataSize
+
+      }
+
+   // create one RSSItem for all cold blocks
```
Why create an RSSItem for cold blocks if there's going to be an RSS item for the cold code in `OMRCodeCache`?

gita-omr replied:
`OMRCodeCache` contains an `RSSRegion` which represents the whole cold part of that cache. `RSSItem` describes some part of the region. There can be many `RSSItem`s in one region. For example, cold code cache can contain cold blocks, snippets, OOL code. I tried to describe that in `OMRRSSReport.hpp`.

dsouzai reviewed (changes requested):
Requested changes + had some questions. Also, can you add a document outlining what an RSSItem is and looks like, and also what the output report is expected to look like?

gita-omr commented on compiler/runtime/OMRRSSReport.cpp:
```diff
@@ -0,0 +1,354 @@
+   addToListSorted(&_pageMap[offset], item);
+
+   // split across pages if necessary
+   size_t tillPageEnd = _pageSize - ((uintptr_t)address % _pageSize);
```
RSSItem can span across multiple pages since it describes an arbitrary range of memory within RSSReport. As in the example I gave below: cold cache region can contain many items: cold blocks, snippets, OOL. Each of then can be bigger than a page. On the other hand, the system reports RSS status per page. That's why we have an array pf pages and each page has a list of items.

dsouzai replied:
Is this a correct understanding:

`RSSItem` can span across multiple pages when it is instantiated, but once you call `addRSSItem`, the `RSSItem` is only going to span a page.

gita-omr replied:
Correct, essentially we will split the item provided by the consumer into multiple items. So the original one will be modified.

gita-omr commented on compiler/runtime/OMRCodeCache.cpp:
```diff
@@ -421,6 +421,13 @@ OMR::CodeCache::initialize(TR::CodeCacheManager *manager,
+   if (OMR::RSSReport::instance())
+      {
+      // TODO: specify page size for the region (vs. the default 4K)
+      _coldRSSRegion = OMR::RSSRegion("cold cache", _coldCodeAlloc, 0, OMR::RSSRegion::highToLow);
```
I am pretty sure it should call the defined constructor with the default values, but I can double check.

gita-omr replied:
Double checked: `OMR::CodeCache` constructor calls defined `OMR::RSSRegion` constructor with the default values.

gita-omr commented on compiler/runtime/OMRRSSReport.cpp:
```diff
@@ -0,0 +1,354 @@
+
+   OMR::RSSItem *newItem = item;
+
+   while (newItem->_size > tillPageEnd)
+      {
+      size_t newSize = newItem->_size - tillPageEnd;
+      newItem->_size = tillPageEnd;
+      address += newItem->_size;
+
+      newItem = new (TR::Compiler->persistentMemory()) OMR::RSSItem(item->_type, address, newSize, counters);
+      offset = (_grows == lowToHigh) ? ++offset : --offset;
+
+      TR_ASSERT_FATAL(offset >= 0, "Got negative offset %d for addr=%p size=%zu type=%s\n", offset,
+                      item->_addr, item->_size, OMR::RSSItem::itemNames[item->_type]);
+
+#ifdef DEBUG_RSS_REPORT
+      TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS adding overflow item addr=%p size=%zu type=%s thread=%d method=%s",
+                                     newItem->_addr, newItem->_size,
+                                     OMR::RSSItem::itemNames[newItem->_type], threadId,
+                                     methodName ? methodName : "no method");
+#endif
+      addToListSorted(&_pageMap[offset], newItem);
+       tillPageEnd = _pageSize;
+      }
```
Actually, I think this code produces least number of checks: we add original item, then check if it was too big, then correct it's size and create/add a new one. Maybe I'll just add a comment clarifying that?

gita-omr replied:
Please note that the first(original) item  is different from all subsequent items. I think the code above is the most concise and avoids separate checking and size calculations for the first item.

dsouzai replied:
Given that the RSSItems are always page sized (except for possibly the last (including the only) one), although something like this:
```
   size_t remainingSize = item->_size;
   size_t tillPageEnd = _pageSize - ((uintptr_t)address % _pageSize);
   
   if (remainingSize > _pageSize)
      item->_size = _pageSize;
   
   remainingSize -= item->_size
      
   addToListSorted(&_pageMap[offset], item);
   
   while (remainingSize)
      {
      size_t newSize = (remainingSize > _pageSize) ? _pageSize : tillPageEnd;
      
      address += newSize;
      OMR::RSSItem *newItem = new (TR::Compiler->persistentMemory()) OMR::RSSItem(item->_type, address, newSize, counters);
      offset = (_grows == lowToHigh) ? ++offset : --offset;
      
      TR_ASSERT_FATAL(offset >= 0, "Got negative offset %d for addr=%p size=%zu type=%s\n", offset,
                         item->_addr, item->_size, OMR::RSSItem::itemNames[item->_type]);
      
   #ifdef DEBUG_RSS_REPORT
         TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS adding overflow item addr=%p size=%zu type=%s thread=%d method=%s",
                                        newItem->_addr, newItem->_size,
                                        OMR::RSSItem::itemNames[newItem->_type], threadId,
                                        methodName ? methodName : "no method");
   #endif
   
      addToListSorted(&_pageMap[offset], newItem);
      
      remainingSize -= newSize;
      }
```
incurs something like one additional comparison operation + one additional subtraction operation per iteration, it is clear what the size of the items are right when they are constructed, and once inserted into the list it isn't mutated. `newItem` is also always a new item.

Essentially, unless this code path is shown to be an extremely hot path that needs every drop of performance (which is unlikely given that this is for reporting and analysis), I don't think the post-insertion size fixup is worth the readability cost.

gita-omr replied:
Sure, I wanted to avoid those extra checks in the code. But perhaps your code is more general in case we change List implementation and start adding items vs pointers to them. I will rewrite.

gita-omr replied:
The above loop is not correct. For example, `address += newSize; `before creating the new item would be wrong since `address` would point to the end of the new item and not the beginning. To be honest, I think the original loop was better. There is nothing wrong with modifying previously inserted item if we know that we inserted a pointer to it. Just couple of comments would make the approach clear: we insert an item and as long as it does not fit into a page we continue creating new items while modifying previous one.

gita-omr replied:
Improved the old version to make it more clear and also add a separate commit with a more sequential version. So we can choose now :)

dsouzai replied:
Ah I see, I didn't realize that the first item can also be less than a page size (unless it was page aligned to begin with).

I mean my big issue was the shallow copy of the counters, which has since been fixed, so with the added comments in commit `f30ce76fd6` I'm less opposed to the original approach; at the very least it's much easier to understand now.

gita-omr commented on compiler/runtime/OMRRSSReport.hpp:
```diff
@@ -0,0 +1,253 @@
+   enum Type _type;
+   uint8_t *_addr;
+   size_t _size;
+   TR_PersistentList<TR::DebugCounterAggregation> _counters;
```
I actually thought that conceptually it works well:` RSSReport` is a global singleton (created in the persistent memory) that contains all the necessary data within itself or points to some persistent data. The TR_Persistent  data structures within it (list of regions, items and so on) hold the same assumption. I think the problem with the copy constructor for the TR_PersistentList within TR RSSItem will only arise if we start copying items. But we don't do it and if we do we actually might want to have a shallow copy (depends on why we would copy items).

dsouzai replied:
I think that in general everything is conceptually well defined; however, assuming that the `TR_PersistentList` won't accidentally be used or changed because `RSSReport` is a singleton is where I believe the conceptual integrity weakens. That coupling of assumptions shouldn't exist. All it would take, for example, is for someone to add a destructor that frees up memory to `TR_PersistentList` and this code would break.

Given that you said that `RSSReport` is a singleton, then perhaps the way around this is to have `RSSReport` have a list of `TR_PersistentList`s (where the first entry is a shared empty list). Then all you have to do, in say `OMR::X86::CodeGenerator::addItemsToRSSReport`, is something like

```
auto coldBlockCountersList = OMR::RSSReport::allocateNewPersistentList();
...
rssItem = new (self()->comp()->trPersistentMemory()) OMR::RSSItem(OMR::RSSItem::coldBlocks, coldCode, actualColdLength,
                                                                  coldBlockCountersList);
rssRegion->addRSSItem(rssItem, codeCache->getReservingCompThreadID(), methodName);
```
where `RSSReport::allocateNewPersistentList` may have a signature such as
```
TR_PersistentList<TR::DebugCounterAggregation>& RSSReport::allocateNewPersistentList();
```

This way the lifetimes of the persistent lists are explicitly tied to the `RSSReport` singleton, and there's no passing by value potential problems.

gita-omr replied:
I was actually thinking we I can improve the following:

1. **Encapsulation**: let OMR::RSSItem handle the storing of the debug counters and have a public method OMR::RSSItem::addDebugCounter(TR::DebugCounterAggregation *counterAggregation)
2. **Memory consumption:** not all items will have counters associated with them. Allocate counters on demand by having only a pointer in the item.

This will also make it very similar to other classes where we check if some list is null and allocate it if needed. However, generally, I would encourage using instance members more while providing all the necessary constructors and operators to the class. I think it's in the spirit of C++ to treat objects as values.

gita-omr replied:
Change instance variable to a pointer (for the counters) to save memory since some items will not have debug counters and we might have a lot of counters. Decided not to encapsulate the list since it makes it much more convenient to pass it as an argument when adding new item.

gita-omr commented:
Addressed comments above.

gita-omr commented:
Sorry, did not create a document yet. I will after we finalize the design in this PR.

gita-omr commented:
Addressed comments and rebased

gita-omr commented:
Made loop in `addRSSItem()` more sequential to avoid an extra store of the item size into memory.

gita-omr commented:
Fixed white spaces.

dsouzai reviewed (changes requested):
Thanks for reworking the persistent counters Gita. Overall LGTM, just had one question about the overlapping regions, and also I'll leave it to you if you want to revert the while loop to the original implementation (with the newly added comments)

gita-omr commented on compiler/runtime/OMRRSSReport.cpp:
```diff
@@ -0,0 +1,362 @@
+      uint8_t *nextStart = nextItem->_addr;
+      uint8_t *nextEnd = nextStart + nextItem->_size;
+
+      if (itemStart < nextEnd || itemEnd > nextStart)
```
Yeah, once again I was trying not to have any special case for the first iteration :)
Due to the way we inserted we can have two situations:
```
1)
   +------------+ 
   |            | 
   +------------+ 
nextStart      nextEnd

        +------------+ 
        |            | 
        +------------+ 
      itemStart     itemEnd 

================================
2)
        +------------+
        |            |
        +------------+
      nextStart    nextEnd

   +------------+ 
   |            | 
   +------------+ 
 itemStart    itemEnd
```
But it's true that only (2) can happen after the first iteration.

gita-omr replied:
The reason we inserted like that is so that any existing overlaping item will be next in the list - to make removal easier.

gita-omr replied:
I added a comment to that point.

dsouzai replied:
I guess I was asking about in the case when it's not overlapping.

Let's just say you have a RSS Item that's page aligned and a page in size. If you have the situation where 
```
+------------+           +------------+
|            |    ->     |            |
+------------+           +------------+
^            ^           ^            ^
ItemStart    ItemEnd     nextStart    nextEnd
```
then won't you remove the item following it? Essentially you added it as per
```
      if (itemStart < nextEnd)
         {
         newElement = itemList->addAfter(item, prevElement);
         break;
         }
```
and then you're removing it based on
```
      if (itemStart < nextEnd || itemEnd > nextStart)
         {
         itemList->removeNext(newElement);
         }
```
so if they don't overlap, won't that still result in the next item getting removed?

gita-omr replied:
Oh, I think you are right! To cover the case (1) above we should have :

`if (itemStart >= nextStart || itemEnd > nextStart)`

gita-omr replied:
Done

0xdaryl commented on compiler/runtime/OMRRSSReport.hpp:
```diff
@@ -0,0 +1,262 @@
+    */
+   TR_PersistentList<RSSItem> getRSSItemList(uint8_t *address)
+         {
+         TR_ASSERT_FATAL(_pageSize >=0, "Page size should be set");
```
Should be `_pageSize > 0`, otherwise it won't catch the case where the page size wasn't set (default is 0).

gita-omr replied:
Thanks for noticing. Of course, that was the whole idea!

gita-omr replied:
Fixed.

0xdaryl commented on compiler/runtime/OMRCodeCache.cpp:
```diff
@@ -421,6 +421,13 @@ OMR::CodeCache::initialize(TR::CodeCacheManager *manager,
 
+   if (OMR::RSSReport::instance())
+      {
+      // Down-stream project should set page size
```
I'm not clear why the downstream project has to set the page size.  Does the `OMR::CodeCache` not know (or could find out) the page size for the code cache memory it is representing?

gita-omr replied:
I think down-stream project can have it's own port library and has its own way of finding the page size. So I thought it's better to leave it to a specific project.

gita-omr replied:
I think it's OK to leave that comment here but also document in RSSReport.hpp

gita-omr replied:
Reworked.

0xdaryl commented on compiler/ras/DebugCounter.cpp:
```diff
@@ -462,6 +463,33 @@ void TR::DebugCounterAggregation::accumulate()
+      TR::DebugCounter *counter = counterDelta->counter;
+      int64_t count = counter->getCount();
+      if (count)
+         TR_VerboseLog::writeLineLocked(TR_Vlog_PERF, "RSS counter count=%d %s", count, counter->getName());
```
"RSS counter" : can a more generic message be printed such that these functions are not tied to the RSS report?

gita-omr replied:
Fixed.

gita-omr commented on compiler/x/codegen/OMRCodeGenerator.cpp:
```diff
@@ -1895,6 +1896,81 @@ struct DescendingSortX86DataSnippetByDataSize
    };
+
+
+void OMR::X86::CodeGenerator::addItemsToRSSReport(uint8_t *coldCode)
```
I would disagree. `coldCode` name is used by the caller. `startOfColdCodeCacheAddress` seems a bit too long to me and I don't think is typical for our source. Do you think the documentation in the header would be sufficient?

gita-omr replied:
Yes, forgot to document.

gita-omr replied:
Documented

gita-omr commented on compiler/ras/DebugCounter.cpp:
```diff
@@ -462,6 +463,33 @@ void TR::DebugCounterAggregation::accumulate()
+      {
+      TR::DebugCounter *counter = counterDelta->counter;
+      int64_t count = counter->getCount();
+      if (count)
```
It's still an experimental feature. Right now it seems like printing zero counters is not useful.

gita-omr replied:
I think I will add an option - `bool printZeroCounters`

gita-omr replied:
Done

gita-omr commented on compiler/runtime/OMRCodeCache.cpp:
```diff
@@ -421,6 +421,13 @@ OMR::CodeCache::initialize(TR::CodeCacheManager *manager,
 
+   if (OMR::RSSReport::instance())
+      {
+      // Down-stream project should set page size
```
I think it's OK to have it here and in the header. Added comment to the constructor.

gita-omr commented on compiler/x/codegen/OMRCodeGenerator.cpp:
```diff
@@ -1895,6 +1896,81 @@ struct DescendingSortX86DataSnippetByDataSize
+
+      TR::Block *block;
+
+      block = node->getBlock();
```
Done.

gita-omr commented on compiler/x/codegen/OMRCodeGenerator.cpp:
```diff
@@ -1895,6 +1896,81 @@ struct DescendingSortX86DataSnippetByDataSize
+
+
+void OMR::X86::CodeGenerator::addItemsToRSSReport(uint8_t *coldCode)
+   {
```
Done

gita-omr commented on compiler/runtime/OMRRSSReport.cpp:
```diff
@@ -0,0 +1,362 @@
+   TR_PersistentList<TR::DebugCounterAggregation> *counters = item->_counters;
+
+   TR_ASSERT_FATAL(address, "Address should not be null");
+   TR_ASSERT_FATAL(_pageSize >=0, "Page size should be set");
```
Fixed!

gita-omr commented:
Addressed latest comments.

gita-omr commented:
I see a problem with setting page size after adding a region - printout can happen in between. I will make page parameter mandatory for the constructor. Making WIP for now.

gita-omr commented:
Removed default RSSRegion constructor to make sure all instance variables are properly initialized as soon as region is created. Region can be created by any component that can provide all the required parameters, including page size.

gita-omr commented:
> Requested changes + had some questions. Also, can you add a document outlining what an RSSItem is and looks like, and also what the output report is expected to look like?

Yes, will add it after the review is close to done.

gita-omr commented:
Rebased.

gita-omr commented:
I think I addressed all the comments and I am not planning on adding any new interfaces to that class right now. @dsouzai @0xdaryl could you please take another look?

dsouzai reviewed (approved):
LGTM

gita-omr commented:
I still need to write a brief document for it.

gita-omr commented:
Small correction in the doc.

dsouzai commented:
Doc looks ok to me; after Daryl's taken a look, could you squash the extra "fix" commits?

0xdaryl commented on compiler/runtime/OMRRSSReport.cpp:
```diff
@@ -0,0 +1,377 @@
+      {
+      if ((regionNum * REGION_PRINT_WIDTH + 1) < MAX_RSS_LINE)
+         {
+         sprintf(rssLine + regionNum * REGION_PRINT_WIDTH, "%*.*s", REGION_PRINT_WIDTH, REGION_PRINT_WIDTH, rssRegion->_name);
```
`sprintf` has been causing grief with newer build compilers due to potential overflow issues being flagged as errors.  I think it's best not to introduce new uses and just go with `snprintf` instead.

See #6863 and a comment from Devin on some ways of consolidating snprintf behaviour across compilers: https://github.com/eclipse/omr/issues/6863#issuecomment-1382937072

gita-omr replied:
Replaced sprintf with TR::snprintfTrunc.

0xdaryl commented on compiler/runtime/OMRRSSReport.cpp:
```diff
@@ -0,0 +1,377 @@
+      {
+      if ((regionNum * REGION_PRINT_WIDTH + 1) < MAX_RSS_LINE)
+         {
+         sprintf(rssLine + regionNum * REGION_PRINT_WIDTH, "%*p", REGION_PRINT_WIDTH, rssRegion->_start);
```
snprintf, see earlier comment.

gita-omr replied:
Replaced sprintf with TR::snprintfTrunc.

0xdaryl commented on compiler/runtime/OMRRSSReport.cpp:
```diff
@@ -0,0 +1,377 @@
+
+      if ((regionNum * REGION_PRINT_WIDTH + 1) < MAX_RSS_LINE)
+         {
+         sprintf(rssLine + regionNum * REGION_PRINT_WIDTH, " %5zu/%5zu(%3.0f%%)", rssSizeKb, sizeKb, (float)rssSizeKb*100.0/sizeKb);
```
snprintf, see earlier comment.

gita-omr replied:
Replaced sprintf with TR::snprintfTrunc.

0xdaryl commented on compiler/runtime/OMRRSSReport.cpp:
```diff
@@ -0,0 +1,377 @@
+         if ((pageCount * PAGE_PRINT_WIDTH + PAGE_PRINT_WIDTH + 1) < MAX_RSS_LINE)
+            {
+            // Note: TR_VerboseLog will truncate the line to some maximum
+            sprintf(pagesInfo + pageCount * PAGE_PRINT_WIDTH, "%1zu ", resident);
```
snprintf, see earlier comment.

gita-omr replied:
Replaced sprintf with TR::snprintfTrunc.

gita-omr commented on compiler/runtime/OMRRSSReport.cpp:
```diff
@@ -0,0 +1,377 @@
+   }
+
+#ifdef LINUX
+typedef struct __attribute__ ((__packed__))
```
The document does not mention `packed`. Can we assume all compilers support it? Otherwise, I am not sure how to check all of them. @0xdaryl

0xdaryl replied:
Presumably, that attribute is required for efficiency rather than correctness.  If so, then I think guarding the typedef using `#ifdef` compiler macros just like in the example here should suffice:
https://github.com/eclipse/omr/blob/d58019eba46edbd5f2e1963e687a27c1dbaccac0/compiler/infra/String.hpp#L31
The `#else` part could just be a typedef without the attribute.  However, if the attribute is required for correctness for some reason then put an `#error` in the `#else` part instead.

gita-omr replied:
The attribute is needed for correctness. If some compiler does not support it I think it will give an error. Also, I noticed this attribute is being used in openj9 without any compiler check.

0xdaryl replied:
OK, I don't think there are any non-Clang or non-gcc based compilers building and testing OMR.  The reference to xlC 12 in the supported compilers list is a bit dated, and we actually plan on revisiting that list at the next OMR architecture meeting.  Newer xlC's are based on Clang, I believe.  So I'll let this go.

gita-omr commented:
Replaced `sprintf `with `TR::snprintfTrunc`.  Since we are printing RSS info in a tabular format (the title  and then rows by rows) this provided an additional benefit of ensuring each field does not exceed its intended length.

gita-omr commented:
Rebased

gita-omr commented:
Corrected and improved latest commit.

gita-omr commented:
> squash

Sure. I have not squashed yet since several new features are waiting for this one and it's easier to rebase those branches when this branch is not squashed.

0xdaryl reviewed (approved):
Thanks for fixing the sprintfs.

gita-omr commented:
Squashed all the commits.

dsouzai commented:
jenkins build all

keithc-ca commented on compiler/runtime/CMakeLists.txt:
```diff
@@ -27,4 +27,5 @@ compiler_library(runtime
 	${CMAKE_CURRENT_LIST_DIR}/OMRCodeCacheManager.cpp
 	${CMAKE_CURRENT_LIST_DIR}/OMRCodeCacheMemorySegment.cpp
 	${CMAKE_CURRENT_LIST_DIR}/OMRCodeCacheConfig.cpp
+	${CMAKE_CURRENT_LIST_DIR}/OMRRSSReport.cpp
```
Shouldn't an equivalent change be made in makefiles like `jitbuilder/build/files/common.mk` for example?

keithc-ca replied:
I suspect this is the reason internal builds are failing (e.g. z/OS):
```
 IEW2456E 9207 SYMBOL addRSSItem__Q2_3OMR9RSSRegionFPQ2_3OMR7RSSItemiPCc        
          UNRESOLVED.  MEMBER COULD NOT BE INCLUDED FROM THE DESIGNATED CALL    
          LIBRARY.                                                              
 IEW2456E 9207 SYMBOL _instance__Q2_3OMR9RSSReport UNRESOLVED.  MEMBER COULD NOT
          BE INCLUDED FROM THE DESIGNATED CALL LIBRARY. NAME SPACE = 3          
FSUM3065 The LINKEDIT step ended with return code 8.
/j9tmp/bld_74387/bld_zos_390/compiler/../compiler/build/rules/zos-xlc/common.mk:47: recipe for target '/j9tmp/bld_74387/bld_zos_390/compiler/../libj9jit29.so' failed
```

dsouzai replied:
@gita-omr ^
